### DISCUSSION TITLE
Feedback as a first time Mill user on Windows

### DISCUSSION THREAD

**Discussion by kevinwu1** [com-lihaoyi/mill#3872] [General]
Hi, just wanted to give some feedback about my experience/troubles setting up mill for the first time.

1. I'm working behind a corporate firewall that blocks repo1.maven.org. I thought that the repositoriesTask would solve this problem, but I actually had to go configure Coursier's mirror.properties file to just get mill running by itself. I'm guessing mill needs to download some stuff before it can actually parse the build.sc file? It took me a really long time to figure this out, especially because on windows, the path for the config file is NOT directly in the home directory. It was C:\Users\MYUSER\AppData\Roaming\Coursier\config\mirror.properties, and C:\Users\MYUSER\AppData\Roaming\Coursier\config\credentials.properties
These paths are not even listed on coursier's documentation

2. The "Installation and IDE Support" page claims that I just need to install the intellij scala plugin and open my project containing a mill build file and intellij will recognize it as a BSP project. However, it actually did not recognize it for me when I first tried. I had to run "mill mill.bsp.BSP/install" before opening it so that intellij can recognize the project as BSP.

3. I've noticed that the documentation on the mill-build website is different for different versions, specifically, the latest 0.12.1 is missing some information found on previous versions.

4. Before I figured out all of the problems, when I ran mill foo.compile, I would get an error message about "Failed to load compiler bridge for 2.12.17 org.scala-lang". I tried switching around the scala version in build.sc, but got the same error for 2.13.11 and 3.3.3. Mysteriously though, 2.13.15 worked. After I figured out the other issues though, I no longer get this error for any scala version.

**Comment by lihaoyi** [comment]
Thanks @kevinwu1!

> 1. I'm working behind a corporate firewall that blocks repo1.maven.org. I thought that the repositoriesTask would solve this problem, but I actually had to go configure Coursier's mirror.properties file to just get mill running by itself. I'm guessing mill needs to download some stuff before it can actually parse the build.sc file? It took me a really long time to figure this out, especially because on windows, the path for the config file is NOT directly in the home directory. It was C:\Users\MYUSER\AppData\Roaming\Coursier\config\mirror.properties, and C:\Users\MYUSER\AppData\Roaming\Coursier\config\credentials.properties
These paths are not even listed on coursier's documentation

I don't have a corporate wifi to test with, but the current documentation recommends you set an environment variable when running Mill:

* https://mill-build.org/mill/0.12.1/javalib/dependencies.html#_working_without_access_to_maven_central

Can you check if that works for you? If it works and the problem is the docs were too hidden, I can try to add a link somewhere in the intro to make it more discoverable

> 2. The "Installation and IDE Support" page claims that I just need to install the intellij scala plugin and open my project containing a mill build file and intellij will recognize it as a BSP project. However, it actually did not recognize it for me when I first tried. I had to run "mill mill.bsp.BSP/install" before opening it so that intellij can recognize the project as BSP.

Can I check what version of IntelliJ and IntelliJ-Scala-Plugin you are using? I've never had to do this, but not sure if it's because I'm on a nightly or EAP build of one of those things

> 3. I've noticed that the documentation on the mill-build website is different for different versions, specifically, the latest 0.12.1 is missing some information found on previous versions.

Yes we have reorganized the docs. In theory everything should still be there, is there anything specific you found missing?

> 4. Before I figured out all of the problems, when I ran mill foo.compile, I would get an error message about "Failed to load compiler bridge for 2.12.17 org.scala-lang". I tried switching around the scala version in build.sc, but got the same error for 2.13.11 and 3.3.3. Mysteriously though, 2.13.15 worked. After I figured out the other issues though, I no longer get this error for any scala version.

I suspect these may be related to the corporate firewall problems, as that would make it impossible to download this necessary component from Maven Central. The fact that it went away after you resolved the other problems does make me lean towards that theory

**Reply by kevinwu1** [reply]
Maybe i'm not doing it right, but here's what I've tested

1. `set COURSIER_REPOSITORIES="https://mycompany.jfrog.io/mycompany/central-maven-remote" & set COURSIER_CREDENTIALS="https://mycompany.jfrog.io/mycompany/central-maven-remote myusername@myemail.com:mypassword" & mill foo.compile`
This results in mill.scalalib.ZincWorkerModule.classpath Failed to load dependencies 
forbidden: https://repo1.maven.org/maven2/org/furesource/jansi/jansi/2.4.1/jansi-2.4.1.jar
2. `set COURSIER_REPOSITORIES=https://mycompany.jfrog.io/mycompany/central-maven-remote & set COURSIER_CREDENTIALS=https://mycompany.jfrog.io/mycompany/central-maven-remote myusername@myemail.com:mypassword & mill foo.compile`
(no quotes this time)
This results in mill.scalalib.ZincWorkerModule.classpath Failed to load dependencies 
forbidden: https://repo1.maven.org/maven2/org/furesource/jansi/jansi/2.4.1/jansi-2.4.1.jar
3. I create a `mirror.properties` and `credentials.properties` files under C:\Users\MYUSER\AppData\Roaming\Coursier\config\
Then I run `mill foo.compile`
This results in 
com.lihaoyi:mill-scalalib-worker_2.13:0.12.1
not found: C:\Users\myuser\.ivy2\local\com.lihaoyi\mill-scalalib-worker_2.13\0.12.1\ivys\ivy.xml
unauthorized: https://mycompany.jfrog.io/mycompany/central-maven-remote/com.lihaoyi\mill-scalalib-worker_2.13 ...
so it looks like it picked up the mirror.properties file but not the credentials.properties file.
Then I run `coursier fetch com.lihaoyi:mill-scalalib-worker_2.13:0.12.1`
This succeeds and the dependency is becomes cached on my local disk, so then mill can get past this dependency.
But then mill runs into unauthorized for another dependency, and i have to manually pull that one in with coursier again (repeat a couple times)


About my intellij versions, im on 2024.2.4 Ultimate Edition Build #IU-242.23726.103
And my scala plugin is on version 2024.2.29

About the docs, the 0.11.12 Scala Installation and IDE support page under the "Debugging IDE issues" section, theres the command `mill --debug mill.bsp/BSP/install`

**Reply by lihaoyi** [reply]
Thanks @kevinwu1!

@alexarchambault do you know what might be going wrong with the `COURSIER_*` environment variables not working?

**Comment by chikei** [comment]
Maybe we can add extra section in getting started or installation to explicitly state what to do if one needs to use maven mirror at the very beginning. Current documentation "we managed dependencies with coursier and repositoriesTask tells coursier what to do. And mill requires some runtime dep is also managed with coursier, but it requires directly instructing coursier" is a bit too reference-y that require a bit effort to carefully read through and some understanding of mill to connect all the points IMO.

**Reply by lihaoyi** [reply]
https://github.com/com-lihaoyi/mill/pull/3881 should move the relevant doc snippets earlier, still need to figure out why the env vars weren't working
